### PR TITLE
Set category title flexibly

### PR DESCRIPTION
**Pull request by myokoym**
TODO:
- Add validation
- Add test

### CODE REVIEW

File: app/models/site.rb [hunk 1]
```diff
@@ -20,4 +21,11 @@ class Site < ActiveRecord::Base
   def credit_enabled?
     participants.exists? && credit_roles.exists?
   end
+
+  def validate_category_title_format
```
Comment (tricknotes) [inline app/models/site.rb:29]: This method is enough to be a private method?

**Comment by okkez** [comment]
Please add more tests.

I think this feature needs unit tests for `escaped_category_title_format` and `validate_category_title_format`.
Because these methods are not so simple.

File: app/models/site.rb [hunk 2]
```diff
@@ -20,4 +21,11 @@ class Site < ActiveRecord::Base
...
+    if /%{(?!name)}/ =~ category_title_format ||
```
Comment (tricknotes) [inline app/models/site.rb]: `category_title_format` may be `nil`?

File: app/models/site.rb [hunk 3]
```diff
@@ -20,4 +21,11 @@ class Site < ActiveRecord::Base
...
+         /%[^{]/ =~ category_title_format
+      errors.add(:category_title_format, "Invalid format")
```
Comment (tricknotes) [inline app/models/site.rb]: Don't assign message string directly.
Please use symbol here to allow using i18n.

File: db/seeds/development/sites.seeds.rb [hunk 4]
```diff
@@ -1,6 +1,7 @@
 Site.create!(
   name: "site1",
   fqdn: ENV["HEROKU_APP_NAME"].present? ? "#{ENV['HEROKU_APP_NAME']}.herokuapp.com" : "localhost",
+  category_title_format: "title (%{name})",
```
Comment (tricknotes) [inline db/seeds/development/sites.seeds.rb]: This seems a optional value.
So can you consider using factory_girls' trait if necessary?

File: test/integration/current_category_test.rb [hunk 5]
```diff
@@ -7,9 +7,9 @@ class CurrentCategoryTest < ActionDispatch::IntegrationTest
   setup do
     travel_to Time.zone.parse("2000-01-01 00:00:00")
 
-    site = Site.create!(name: "daimon-news", fqdn: "www.example.com")
+    @site = Site.create!(name: "daimon-news", fqdn: "www.example.com")
```
Comment (tricknotes) [inline test/integration/current_category_test.rb:10]: This shouldn't depend default factory data. Because it is so fragile.
If you want to assert `Site#category_title_format`, you can specify it in here.

---

そもそも `Site#category_title_format` のアサートを他のテストを一緒くたにしないほうがいいのでは…？

File: app/decorators/category_decorator.rb [hunk 6]
```diff
@@ -20,10 +20,19 @@ def canonical_url(posts)
   end
 
   def page_title(posts)
+    if site.category_title_format.present?
+      begin
+        title = site.category_title_format % {category_name: name}
+      rescue KeyError
```
Comment (tricknotes) [inline app/decorators/category_decorator.rb]: I think this should be validated on date changed.

Comment (tricknotes) [inline app/decorators/category_decorator.rb]: Ah right. the validation is already set.
In this case, is `rescue` unnecessary?

Comment (okkez) [inline app/decorators/category_decorator.rb]: I added code that restrict `Site#category_title_format` in 747edba.
This is the safe net if my implementation of `Site#validate_category_title_format` have bugs.
This code improves robustness.

Comment (tricknotes) [inline app/decorators/category_decorator.rb]: > This code improves robustness.

I don't think so.
This `rescue` will hide some bugs against our expectation.

If there are some bugs, it should be raised to notify our bug tracking system.

Comment (okkez) [inline app/decorators/category_decorator.rb]: I will add `logger.error` for this.

File: app/views/admin/sites/_form.html.haml [hunk 7]
```diff
@@ -5,6 +5,7 @@
   = f.url_field :css_url
   = f.text_field :fqdn
   = f.text_field :tagline
+  = f.text_field :category_title_format
```
Comment (tricknotes) [inline app/views/admin/sites/_form.html.haml]: How do you think about help text?
The user can't recognize what `%{category_name}` is used as placeholder.

File: test/models/site_test.rb [hunk 8]
```diff
@@ -61,4 +61,35 @@ class SiteTest < ActiveSupport::TestCase
...
     end
   end
+
+  sub_test_case "category_title_format" do
+    test "null" do
+      site = build(:site)
+      assert_true(site.valid?)
+    end
+
+    test "blank" do
```
Comment (tricknotes) [inline test/models/site_test.rb:71]: It is intentional that the assertion for "blank" is not included to the following `data`?

File: test/models/site_test.rb [hunk 9]
```diff
@@ -61,4 +61,35 @@ class SiteTest < ActiveSupport::TestCase
       assert_equal(["はすでに存在します"], site.errors[:fqdn])
     end
   end
+
+  sub_test_case "category_title_format" do
+    test "null" do
```
Comment (tricknotes) [inline test/models/site_test.rb:66]: ref: https://github.com/bm-sms/daimon-news/pull/458#discussion_r59831878
also this.

File: app/views/admin/sites/_form.html.haml (hunk 7 above)
Comment (okkez) [inline app/views/admin/sites/_form.html.haml]: Good catch!
I want to add help text.

File: test/models/site_test.rb (hunk 8 above)
Comment (okkez) [inline test/models/site_test.rb:71]: Yes, It's intentional.
"blank" and "null" is special case because user cannot input these data via web UI.
But we should add test for these case, we should consider that "blank" and "null" may break something or not.

Comment (tricknotes) [inline test/models/site_test.rb:71]: I see. It's reasonable.

**Comment by mtsmfm** [comment]
Could you rebase this branch?

**Comment by mtsmfm** [comment]
LGTM other than rebasing

**Comment by okkez** [comment]
OK

**Comment by okkez** [comment]
Rebased

**Comment by mtsmfm** [comment]
:cool:
